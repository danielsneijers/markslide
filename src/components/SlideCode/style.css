.slide {
  height: 100vh;
  width: 100vw;
  color: #FAFAFA;
  padding: 10%;
  display: flex;
  align-items: center;
  justify-content: center;
  overflow: hidden;
  user-select: none;
  background: #1B2B34;
}

.code {
  transition: opacity .2s ease-in-out;
  transform: scale(1) translateY(10px);
}

/* stylelint-disable-next-line selector-class-pattern */
:global .slide-code-appear {
  opacity: .01;
}

/* stylelint-disable-next-line selector-class-pattern */
:global .slide-code-appear.slide-code-appear-active {
  opacity: 1;
  transition: opacity .2s ease-in-out;
}

/* stylelint-disable-next-line selector-class-pattern */
:global .slide-code-enter {
  opacity: .01;
}

/* stylelint-disable-next-line selector-class-pattern */
:global .slide-code-enter.slide-code-enter-active {
  opacity: 1;
  transition: opacity .2s ease-in-out;
}

/* stylelint-disable-next-line selector-class-pattern */
:global .slide-code-leave {
  opacity: 1;
}

/* stylelint-disable-next-line selector-class-pattern */
:global .slide-code-leave.slide-code-leave-active {
  opacity: .01;
  transition: opacity .2s ease-in-out;
}
