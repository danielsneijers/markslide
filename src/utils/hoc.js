// @flow
export const getDisplayName = (WrappedComponent: { [string]: string }) =>
  WrappedComponent.displayName || WrappedComponent.name || 'Component'
