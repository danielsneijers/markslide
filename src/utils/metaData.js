// @flow
import { compose } from 'ramda'

export type SlideWithMetaData = {
  content: string,
  meta: {}
};

export type MetaData = {
  [string]: string
};

export type LocMetaData = {
  [string]: string,
  loc?: Array<Array<number>>
};

const META_REGEX = /{:.*}/g

export const findMetaDataInContent = (slide: string): Array<string> =>
  slide.match(META_REGEX) || []

export const convertMetaDataMatchesToMetaObject = (
  matches: Array<string>
): {} =>
  matches.reduce((acc, match) => {
    const key = match.split(' ')[0]
    const value = match.replace(key, '').slice(0, -1).trim()

    return { ...acc, [key.substring(2)]: value }
  }, {})

export const parseLocFromMetaData = (metaData: MetaData): LocMetaData => {
  if (!metaData.loc) {
    return { ...metaData }
  }

  try {
    return { ...metaData, loc: JSON.parse(`[${metaData.loc}]`) }
  } catch (e) {
    return {}
  }
}

export const extractAndConvertMetaData: Function = compose(
  parseLocFromMetaData,
  convertMetaDataMatchesToMetaObject,
  findMetaDataInContent
)

export const contentWithOutMetaData = (content: string): string => {
  const matches = findMetaDataInContent(content)
  return matches.reduce((acc, match) => {
    return acc.replace(match, '').trim()
  }, content)
}

export const separateContentAndMetaData = (
  slide: string
): SlideWithMetaData => {
  return {
    content: contentWithOutMetaData(slide),
    meta: extractAndConvertMetaData(slide)
  }
}
