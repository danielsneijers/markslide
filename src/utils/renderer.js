// @flow
import Marked from 'marked'
import Prism, { highlight } from 'prismjs'
import { compose } from 'ramda'
import { colors } from 'config/settings'

export function highlightWithPrism (code: string, lang: string): string {
  require('prismjs/plugins/normalize-whitespace/prism-normalize-whitespace')
  const NWPlugin = Prism.plugins.NormalizeWhitespace
  const normalizedCode = NWPlugin.normalize(code, {
    'remove-trailing': false,
    'remove-indent': false,
    'left-trim': false,
    'right-trim': false
  })

  return wrapRowsInSpan(highlight(
    normalizedCode,
    Prism.languages[lang]
  ))
}

export const wrapRowsInSpan = (text: string): string => {
  return text.split(/\n/)
    .map((line) => `<span className='code-row'>${line}</span>\n`)
    .join('')
}

export const enhanceTextWithHeaders = (text: string, level: number = 1): string => {
  return `<h${level}>${text}</h${level}>\n`
}

export const enhanceTextWithParagraph = (text: string): string => {
  return `<p>${text}</p>\n`
}

export const enhanceTextWithColor = (text: string): string => {
  const availableColors = Object.keys(colors)

  return availableColors.reduce((acc, color) => {
    return acc
      .replace(`<${color}>`, `<span style=color:${colors[color]}>`)
      .replace(`</${color}>`, `</span>`)
  }, text)
}

export const enhanceHeaders: string = compose(enhanceTextWithColor, enhanceTextWithHeaders)
export const enhanceParagraphs: string = compose(enhanceTextWithColor, enhanceTextWithParagraph)

const renderer = new Marked.Renderer()
renderer.heading = enhanceHeaders
renderer.paragraph = enhanceParagraphs

Marked.setOptions({
  renderer,
  gfm: true,
  tables: true,
  breaks: true,
  pedantic: false,
  sanitize: false,
  smartLists: true,
  smartypants: false,
  highlight: highlightWithPrism
})

export default renderer
