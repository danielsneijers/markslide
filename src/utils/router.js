import { get } from 'lodash'

export const getFromRouterProps = (props, param) => {
  return get(props, `match.params.${param}`)
}

export const getParamFromRouterProps = (param) => (props) => {
  return getFromRouterProps(props, param)
}

export const getSlideIndexFromProps = getParamFromRouterProps('slide')
