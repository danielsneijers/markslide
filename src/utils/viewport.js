// @flow
import { compose } from 'ramda'
import { memoize } from 'lodash'

type ElementDimensions = {
  width: number,
  height: number
}

export const getElementSize = (element: HTMLElement): ElementDimensions => {
  const { width, height } = element.getBoundingClientRect()

  return {
    width: Math.floor(width),
    height: Math.floor(height)
  }
}

export const memoizedGetElementSize: Function = memoize(getElementSize)

export const dimensionsFitViewport = (elementDimensions: ElementDimensions): boolean => {
  const { width, height } = elementDimensions

  return window.innerWidth >= width && window.innerHeight >= height
}

export const elementOutOfBoundFraction = (elementDimensions: ElementDimensions): ElementDimensions => {
  const { width, height } = elementDimensions

  return {
    width: window.innerWidth / width,
    height: window.innerHeight / height
  }
}

export const scaleToFit = (outOfBoundFraction: ElementDimensions): number => {
  const { width, height } = outOfBoundFraction

  return width > height ? height : width
}

export const elementFitsViewport: Function = compose(dimensionsFitViewport, memoizedGetElementSize)
export const scaleElementToFit: Function = compose(scaleToFit, elementOutOfBoundFraction, memoizedGetElementSize)
