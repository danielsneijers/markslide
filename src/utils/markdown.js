// @flow
import Marked from 'marked'
import { compose } from 'ramda'
import { separateContentAndMetaData } from 'utils/metaData'
import type { SlideWithMetaData } from 'utils/metaData'
import renderer from 'utils/renderer'

type ParsedSlideWithMetaData = {
  content: HTMLElement,
  meta: {}
}

export const splitSlides = (content: string): Array<string> =>
  content.split('\n---\n')

export const slidesCount = (content: string): number =>
  splitSlides(content).length

export const parseSlidesToEnhancedObjects = (
  slides: Array<string>
): Array<SlideWithMetaData> => slides.map(separateContentAndMetaData)

export const parseSlideObjectsToHtml = (
  slides: Array<SlideWithMetaData>
): Array<ParsedSlideWithMetaData> =>
  slides.map(({ content, meta }: SlideWithMetaData) => {
    return { content: Marked(content, { renderer }), meta }
  })

export const parse: Function = compose(
  parseSlideObjectsToHtml,
  parseSlidesToEnhancedObjects,
  splitSlides
)
