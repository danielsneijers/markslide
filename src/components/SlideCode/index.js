// @flow
import React, { PureComponent } from 'react'
import { debounce } from 'lodash'
import { elementFitsViewport, scaleElementToFit } from 'utils/viewport'
import CSS from './style.css'

type Props = {
  uniqueKey: string,
  content: string
}

type State = {
  style: {
    transform?: string,
    position?: string,
    width?: string
  }
}

const SLIDE_ID: string = 'code-slide-content'

class SlideCode extends PureComponent {
  props: Props
  state: State

  state = {
    style: {}
  }

  componentDidMount () {
    this.checkForScaling()

    window.addEventListener('resize', this.checkForScaling)
  }

  componentWillReceiveProps () {
    this.checkForScaling()
  }

  componentWillUnMount () {
    window.removeEventListener('resize', this.checkForScaling)
  }

  checkForScaling: Function = debounce(() => {
    const wrapper = document.getElementById(SLIDE_ID)
    const scaleFraction = elementFitsViewport(wrapper) ? 1 : scaleElementToFit(wrapper)
    const width = 90 / scaleFraction

    const style = {
      transform: `scale(${scaleFraction}) translateY(0)`,
      position: 'absolute',
      width: `${width}%`,
      opacity: 1
    }

    this.setState({ style })
  }, 200)

  render (): React$Element<any> {
    const { content, uniqueKey } = this.props

    return (
      <div className={CSS.slide} key={uniqueKey}>
        <div
          id={SLIDE_ID}
          className={CSS.code}
          style={{ opacity: 0, ...this.state.style }}
          dangerouslySetInnerHTML={{ __html: content }}
        />
      </div>
    )
  }
}

export default SlideCode
