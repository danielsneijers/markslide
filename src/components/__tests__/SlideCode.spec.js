import React from 'react'
import { shallow } from 'enzyme'
import SlideCode from 'components/SlideCode'
import * as viewport from 'utils/viewport'

jest.useFakeTimers()

let tree, instance
const content = `<pre><code class='lang-javascript'>console<span class='token punctuation'>.</span><span class='token function'>log</span><span class='token punctuation'>(</span><span class='token string'>'hello world!'</span><span class='token punctuation'>)</span></code></pre>`

describe('components/SlideCode', () => {
  beforeEach(() => {
    tree = shallow(<SlideCode content={content} />)
    instance = tree.instance()

    window.addEventListener = jest.fn()
  })

  afterEach(() => {
    window.addEventListener.mockReset()
  })

  test('renders correctly', () => {
    expect(tree).toMatchSnapshot()
  })

  test('checks if scaling is needed when component is mounted', () => {
    instance.checkForScaling = jest.fn()
    expect(instance.checkForScaling).toHaveBeenCalledTimes(0)

    instance.componentDidMount()
    expect(instance.checkForScaling).toHaveBeenCalledTimes(1)
  })

  test('adds a window resize listener when component is mounted', () => {
    instance.componentDidMount()

    expect(window.addEventListener).toHaveBeenCalledTimes(1)
    expect(window.addEventListener).toBeCalledWith('resize', instance.checkForScaling)
  })

  test('checks if scaling is needed when component receives props', () => {
    instance.checkForScaling = jest.fn()
    expect(instance.checkForScaling).toHaveBeenCalledTimes(0)

    tree.setProps({ content })
    tree.update()

    expect(instance.checkForScaling).toHaveBeenCalledTimes(1)
  })

  describe('checkForScaling', () => {
    it('doesn\'t apply scaling when content fits viewport', () => {
      const setStateArguements = {
        style: {
          opacity: 1,
          position: 'absolute',
          transform: 'scale(1) translateY(0)',
          width: '90%'
        }
      }

      instance.setState = jest.fn()
      viewport.elementFitsViewport = jest.fn()
      viewport.elementFitsViewport.mockReturnValue(true)

      expect(instance.setState).toHaveBeenCalledTimes(0)

      instance.checkForScaling()
      jest.runAllTimers()

      expect(instance.setState).toHaveBeenCalledTimes(1)
      expect(instance.setState).toBeCalledWith(setStateArguements)
    })
  })

  it('applies scaling when content doesn\'t fit viewport', () => {
    const setStateArguements = {
      style: {
        opacity: 1,
        position: 'absolute',
        transform: 'scale(0.5) translateY(0)',
        width: '180%'
      }
    }

    instance.setState = jest.fn()
    viewport.elementFitsViewport = jest.fn()
    viewport.elementFitsViewport.mockReturnValue(false)
    viewport.scaleElementToFit = jest.fn()
    viewport.scaleElementToFit.mockReturnValue(0.5)

    expect(instance.setState).toHaveBeenCalledTimes(0)

    instance.checkForScaling()
    jest.runAllTimers()

    expect(instance.setState).toHaveBeenCalledTimes(1)
    expect(instance.setState).toBeCalledWith(setStateArguements)
  })
})
