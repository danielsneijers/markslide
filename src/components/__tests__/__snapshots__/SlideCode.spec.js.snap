// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/SlideCode renders correctly 1`] = `
<SlideCode
  content="<pre><code class='lang-javascript'>console<span class='token punctuation'>.</span><span class='token function'>log</span><span class='token punctuation'>(</span><span class='token string'>'hello world!'</span><span class='token punctuation'>)</span></code></pre>"
  index={1}
>
  <div
    className="slide code-slide slide-1"
  >
    <div
      className="code"
      dangerouslySetInnerHTML={
        Object {
          "__html": "<pre><code class='lang-javascript'>console<span class='token punctuation'>.</span><span class='token function'>log</span><span class='token punctuation'>(</span><span class='token string'>'hello world!'</span><span class='token punctuation'>)</span></code></pre>",
        }
      }
      style={
        Object {
          "opacity": 0,
        }
      }
    />
  </div>
</SlideCode>
`;
